### PR TITLE
Provide basic support for `assignment` Task Listeners after `ASSIGN` command

### LINKED ISSUE #24232: Support a basic `assignment` task listener
```[tasklist]
### Tasks
- [ ] Move appending the `ASSIGNED` event to `onFinalizeCommand` for `UserTaskAssignProcessor`
- [ ] Allow multiple assignment listeners, i.e. Track assignment task listener index
- [ ] Store user task intermediate state on `ASSIGNING` event
- [ ] Cleanup user task intermediate state on `ASSIGNED` event
- [ ] Favor the user task intermediate state in the task listeners and finalizeCommand
- [ ] Testing
```

### PR DESCRIPTION
**Pull request by ce-dmelnych**
## Description

This PR is the first of two required to provide foundational support for `assignment` task listeners on the User Task element. Specifically, this PR enables `assignment` task listeners that will be triggered after the `ASSIGN` command, covering both `assign` and `unassign` operations on a User Task.

### Out of scope:
- support for `assignment` task listeners that should be triggered after the `CLAIM` User Task command will be implemented in a separate PR.

## Checklist

- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

## Related issues

relates to #24232

### CODE REVIEW

**Review by ana-vinogradova-camunda — APPROVED** [review]
🚀 Great work, thank you @ce-dmelnych !
I took a second look after our sync meeting today, LGTM 👍 

Thank you for refactoring the tests 🥇

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/UserTaskAssigningV2Applier.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+  public UserTaskAssigningV2Applier(final MutableProcessingState processingState) {
+    userTaskState = processingState.getUserTaskState();
+  }
+
+  @Override
+  public void applyState(final long key, final UserTaskRecord value) {
+    userTaskState.updateUserTaskLifecycleState(key, LifecycleState.ASSIGNING);
+    System.out.println("DMK: Key:" + key);
+    System.out.println("DMK: write intermediate state");
```
Comment (korthout) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/UserTaskAssigningV2Applier.java]: 🔧 Please remember to remove these

✅ EDIT: Thank you 🙌

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/listeners/task/TaskListenerTest.java [hunk 2]
```diff
@@ -163,6 +163,8 @@ public void shouldAssignUserTaskAfterAllAssignmentTaskListenersAreExecuted() {
             Assertions.assertThat(userTask)
                 .hasAssignee("me")
                 .hasAction("my_assign_action")
+                .describedAs(
+                    "Expected all variables provided during UserTask assignment to be ignored")
```
Comment (korthout) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/listeners/task/TaskListenerTest.java:167]: 💭 Yeah, variables aren't support for the assign command. However, it's not exactly related to task listeners but to task assignment in general. So, I'd be fine with removing the variables from this test completely.

**Review by korthout — APPROVED** [review]
Great PR @ce-dmelnych 🚀 

LGTM 👍

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/UserTaskAssignedV2Applier.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  @Override
+  public void applyState(final long key, final UserTaskRecord value) {
+    final String assignee = value.getAssignee();
+    final UserTaskRecord userTask = userTaskState.getUserTask(key);
+    userTask.setAssignee(assignee);
+    userTaskState.update(userTask);
+    userTaskState.updateUserTaskLifecycleState(key, LifecycleState.CREATED);
+    userTaskState.deleteIntermediateState(key);
+    userTaskState.deleteRecordRequestMetadata(key);
+  }
```
[the remark this answers is not in the document]
Comment (ce-dmelnych) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/UserTaskAssignedV2Applier.java:35]: Thanks for the wise suggestion, Nico! 🙌
I’ll follow this approach when we'll work on adding support for  the `update` listeners.

**Pull request merged**